pkg/controller/windowsmachineconfig/tracker.go

```diff
@@ -79,14 +75,110 @@ func newNodeRecord(ipAddress string, secretName string) nodeRecord {
+	}
+	// Set up SSH client
+	if err := winVM.GetSSHClient(); err != nil {
+		return nil, errors.Wrap(err, "error instantiating ssh client")
+	}
+	return winVM, nil
+}
+
+// getInstanceIP gets the instance IP address associated with a node
+func getInstanceIP(nodeList *v1.NodeList, instanceID string) (string, error) {
+	// Ignore the nodes that are not ready.
+	for _, node := range nodeList.Items {
+		for _, condition := range node.Status.Conditions {
+			if condition.Type != v1.NodeReady {
+				continue
+			}
+		}
+		if strings.Contains(node.Spec.ProviderID, instanceID) {
+			for _, address := range node.Status.Addresses {
+				// If external ip exists return it as it is used in AWS, if it doesn't return internal(in case of azure)
+				// TODO: collect all the ips and define a priority order, external then internal etc.
+				if address.Type == v1.NodeExternalIP && len(address.Address) > 0 {
+					return address.Address, nil
+				}
+			}
+		}
+	}
+	return "", errors.New("unable to find Windows Worker nodes")
+}
+
+// getCredsFromSecret gets the credentials associated with the instance.
+func getCredsFromSecret(k8sclientset *kubernetes.Clientset, instanceID string, operatorNS string) (Credentials, error) {
+	creds := Credentials{}
+	instanceSecret, err := k8sclientset.CoreV1().Secrets(operatorNS).Get(instanceID, metav1.GetOptions{})
+	if err != nil && k8serrors.IsNotFound(err) {
+		return creds, errors.Wrap(err, "getting secret")
+	}
+	encodedCreds := instanceSecret.Data[instanceID]
+	if err := json.Unmarshal(encodedCreds, &creds); err != nil {
+		return creds, errors.Wrap(err, "unmarshalling creds failed")
+	}
+	return creds, err
+}
+
+// initWindowsVMs initializes the windowsVMs map from the store configmMap and corresponding secrets
+func initWindowsVMs(k8sclientset *kubernetes.Clientset, operatorNS string) (map[types.WindowsVM]bool, error) {
+	windowsVMs := make(map[types.WindowsVM]bool)
+	store, err := k8sclientset.CoreV1().ConfigMaps(operatorNS).Get(StoreName, metav1.GetOptions{})
+	// It is ok if the tracker configmap doesn't exist. It is usually the case when the operator starts
+	// for the first time but it is not ok when the tracker configmap exists but it unqueryable for some reason
+	if err != nil && !k8serrors.IsNotFound(err) {
+		return nil, errors.Wrap(err, fmt.Sprintf("unable to query %s/%s ConfigMap", operatorNS, StoreName))
+	}
+	if err != nil && k8serrors.IsNotFound(err) {
+		log.Info("tracker config map doesn't exist yet so we can skip initialization of the WindowsVM map")
+		return windowsVMs, nil
+	}
+	// TODO: Get all the running Windows nodes in the cluster
+	//		jira story: https://issues.redhat.com/browse/WINC-280
+	nodeList, err := k8sclientset.CoreV1().Nodes().List(metav1.ListOptions{LabelSelector: nc.WindowsOSLabel})
+	if err != nil {
+		return nil, errors.Wrap(err, "error while querying for Windows nodes")
+	}
+	for instanceID := range store.BinaryData {
```

Comment:
If one of these items doesn't have the correct information they should be removed from the configmap, should they not? Maybe a best effort should also be made to remove VM?

Reply:
No, configmap should be source of truth, if something changes in the configmap, it should be stomped back but that is beyond the scope of this PR.

The flow would be :
`On update or delete of the configmap, we should generate the configmap again(Have a watch for the configmap)
- We should use windowsVM and the list of windows Nodes available in the cluster to get the information.
-  The logic would be something like query the list of Windows nodes we have in cluster and compare them with what we have in the windowsVM map, if they match query the cloud provider api and update the configmap`

Reply:
Please add this as a comment